### DISCUSSION TITLE
Properly configuring TLS

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [kanidm/kanidm#2403] [Q&A] [answered]
I was recently using https://openid.net in an attempt to get some insight on how the authorization flow, and perhaps why this error occurs: https://github.com/9p4/jellyfin-plugin-sso/discussions/156

In the end, I was able to get mostly everything working - with the exception of the JWT token phase - it threw a generic error about being unable to verify the token.

Through my limited knowledge of TLS and JWT/S, I believe this might be due to my current HTTPS setup. I currently have:
```
Client: <-TLS/TRUSTED-> Traefik <-TLS/SELFSIGNED-> Kanidm
```
My best guess is that the JWT used by kanidm for OAUTH2/OpenID  is signed with the self-signed TLS certificate, and the client throws an error since it can't verify the self-signed certificate via the system CA bundle.

Is my guess correct, or is more information needed to properly assess the issue?

Regardless, I would be interested to know - should the certificate used by Kanidm be trusted via the client system, whether via the system CA bundle or a self-signed manually installed public cert? 

Apologies if the question is a little vague, I'm stretching past the limits of my current knowledge.

**Comment by TheRealGramdalf** [comment]
Additionally, it would be nice to know if you can change the certificate post initial installation, and if there are any things of note (similar to running a database vacuum after switching the page size).

**Reply by yaleman** [reply, marked as answer]
JWT's are signed by an internal certificate not the client-facing TLS one. The public key is linked from the OIDC discovery data (`jwk` field),  `https://idm.example.com/oauth2/openid/<oauth_rs_id>/public_key.jwk`.

The JWT signing keys don't need to be publicly trusted certs because if you're verifying them, you'll also be collecting the cert from above as part of the process.

The HTTPS API certificate needs to be trusted, for (hopefully) obvious reasons.

**Reply by Firstyear** [reply]
@yaleman is correct. The TLS keys you provide are only used for HTTPS and LDAPS. Otherwise, Kanidm internally generates signing keys for JWT's and other materials.

As @yaleman said, this means your client *trusts* the HTTPS certificate when they retrieve the public_key.jwk, which then allows the client to trust that the jwk is itself authentic. This jwk can then be used to verify the OIDC tokens. 

Note that each oauth2 client has a unique signing key - this is to prevent the tokens of one client being confused with another, and to allow "failure domains" so that keys of one client can be rolled without affecting others.

**Reply by yaleman** [reply]
And remember - the OAuth2 client is the application you're signing into - not your browser/laptop/whatever.

**Reply by TheRealGramdalf** [reply]
Got it. So in the example given in the OP, the client verifies the server's identity via the HTTPS certificate provided by traefik, and using that established trust it queries the API for the JWT public key via an OAUTH2/OpenID discovery mechanism.

**Reply by yaleman** [reply]
> Is the JWT signed by the TLS certificate specified by the user (which is different than the client facing cert in the above example), or is that managed internally by Kanidm?

It's all managed internally.

> To expand on https://github.com/kanidm/kanidm/discussions/2403#discussioncomment-7975817, I'm referring to changing the cert used for communication between traefik and Kanidm

Change it as often as you like.
